src/node_locks.cc

```diff
@@ -0,0 +1,300 @@
+      held_locks_.push_back(std::move(lock));
+
+      // Let r be the result of invoking callback with a
+      // new Lock object associated with lock as the only argument.
+
+      TryCatch try_catch(env->isolate());
+      Local<Value> r;
+      Local<Object> lockObj = Object::New(env->isolate());
+      lockObj
+          ->Set(env->context(),
+                env->name_string(),
+                String::NewFromUtf8(env->isolate(), lock->name().c_str()))
+          .ToChecked();
+      lockObj
+          ->Set(env->context(),
+                env->mode_string(),
+                lock->mode() == Lock::Mode::kShared ? env->shared_string()
+                                                    : env->exclusive_string())
+          .ToChecked();
+      Local<Value> args[] = { lockObj };
+      if (request->callback()
+              ->Call(env->context(), Undefined(env->isolate()), 1, args)
+              .ToLocal(&r)) {
+        waiting->Resolve(env->context(), r).ToChecked();
+      } else {
+        waiting->Reject(env->context(), try_catch.Exception()).ToChecked();
+      }
+    }
+  }
+}
+
+bool LockManager::IsGrantable(const LockRequest* request) {
+  // If mode is "exclusive", then return true if all of the
+  // following conditions are true, and false otherwise:
+  if (request->mode() == Lock::Mode::kExclusive) {
+    Mutex::ScopedLock held_lock(held_mutex_);
+    // No lock in held has a name that equals name
+    for (auto const& lock : held_locks_) {
+      if (lock->name() == request->name()) {
+        return false;
+      }
+    }
+
+    Mutex::ScopedLock pending_lock(pending_mutex_);
+    // No lock request in queue earlier than request has a name that equals
+    // name.
+    for (auto const& req : pending_requests_) {
+      if (request == req.get()) {
+        break;
+      }
+      if (req->name() == request->name()) {
+        return false;
+      }
+    }
+    return true;
+  } else {
+    // Otherwise, mode is "shared"; return true if all of the
+    // following conditions are true, and false otherwise:
+
+    Mutex::ScopedLock held_lock(held_mutex_);
+    // No lock in held has mode "exclusive" and has a name that equals name.
+    for (auto const& lock : held_locks_) {
+      if (lock->name() == request->name() &&
+          lock->mode() == Lock::Mode::kExclusive) {
+        return false;
+      }
+    }
+
+    Mutex::ScopedLock pending_lock(pending_mutex_);
+    // No lock request in queue earlier than request has a
+    // mode "exclusive" and name that equals name
+    for (auto const& req : pending_requests_) {
+      if (request == req.get()) {
+        break;
+      }
+      if (req->name() == request->name() &&
+          req->mode() == Lock::Mode::kExclusive) {
+        return false;
+      }
+    }
+    return true;
+  }
+}
+
+// promise, callback, name, mode, ifAvailable, steal
+void LockManager::Request(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
```

Comment:
Likely better to push it to a separate follow up PR, but it would likely be helpful to emit trace events to track lock requests, grants, and releases to make them easier to debug.